module/VuFind/src/VuFind/Db/Service/TagService.php

```diff
@@ -71,6 +80,346 @@ public function __construct(
+        $retVal['present'] = array_unique($results);
+
+        // Detect missing IDs:
+        $retVal['missing'] = array_diff($ids, $retVal['present']);
+
+        // Send back the results:
+        return $retVal;
+    }
+
+    /**
+     * Get resources associated with a particular tag.
+     *
+     * @param string $tag  Tag to match
+     * @param string $user ID of user owning favorite list
+     * @param string $list ID of list to retrieve (null for all favorites)
+     *
+     * @return array
+     */
+    public function getResourceIDsForTag($tag, $user, $list = null)
+    {
+        $dql = 'SELECT DISTINCT(rt.resource) AS resource_id '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t '
+            . 'WHERE ' . ($this->caseSensitive ? 't.tag = :tag' : 'LOWER(t.tag) = LOWER(:tag) ')
+            . 'AND rt.user = :user ';
+
+        $user = is_object($user) ? $user : $this->entityManager->getReference(User::class, $user);
+        $parameters = compact('tag', 'user');
+        if (null !== $list) {
+            $list = is_object($list) ? $list : $this->entityManager->getReference(UserList::class, $list);
+            $dql .= 'AND rt.list = :list';
+            $parameters['list'] = $list;
+        }
+        $query = $this->entityManager->createQuery($dql);
+        $query->setParameters($parameters);
+        $result =  $query->getSingleColumnResult();
+        return $result;
+    }
+
+    /**
+     * Get lists associated with a particular tag.
+     *
+     * @param string|array      $tag        Tag to match
+     * @param null|string|array $listId     List ID to retrieve (null for all)
+     * @param bool              $publicOnly Whether to return only public lists
+     * @param bool              $andTags    Use AND operator when filtering by tag.
+     *
+     * @return array
+     */
+    public function getListsForTag(
+        $tag,
+        $listId = null,
+        $publicOnly = true,
+        $andTags = true
+    ) {
+        $tag = (array)$tag;
+        $listId = $listId ? (array)$listId : null;
+        $dql = 'SELECT IDENTITY(rt.list) as list, COUNT(DISTINCT(rt.tag)) AS tagcnt '
+            . 'FROM ' . $this->getEntityClass(ResourceTags::class) . ' rt '
+            . 'JOIN rt.tag t '
+            . 'JOIN rt.list l '
+            . 'WHERE rt.resource IS NULL ';
+        $parameters = [];
+        if (null !== $listId) {
+            $dql .= 'AND rt.list IN (:listId) ';
+            $parameters['listId'] = $listId;
+        }
+        if ($publicOnly) {
+            $dql .= 'AND l.public = :public ';
+            $parameters['public'] = 1;
+        }
+        if ($tag) {
+            if ($this->caseSensitive) {
+                $dql .= 'AND t.tag IN (:tag) ';
+                $parameters['tag'] = $tag;
+            } else {
+                $dql .= 'AND LOWER(t.tag) IN (:tag) ';
+                $parameters['tag'] = array_map('strtolower', $tag);
```

Comment:
The original code is subtly different from this -- it does all the lowercasing in the database, instead of doing some in the database and some in PHP. I think it is better if we can continue to do things that way, because sometimes different regions or locales may have different lowercasing rules, and this could result in inconsistent behavior if the database and PHP apply different algorithms.

I'm not sure what the easiest way to do this would be, but I guess a crude but effective approach might be something like:

```
$tagClauses = [];
foreach ($tag as $i => $currentTag) {
    $tagPlaceholder = 'tag' . $i;
    $tagClauses[] = 'LOWER(t.tag) = LOWER(:' . $tagPlaceholder . ')';
    $parameters[$tagPlaceholder] = $currentTag;
}
$dql .= 'AND (' . implode(' OR ', $tagClauses) . ')';
```

...maybe Doctrine offers some clever way to do this in less code, but if not, I think this is okay.

Reply:
I couldn't find a better way of doing it in Doctrine. I was surprised to know that case, I did some reading and got to know that the conversions often go wrong. The only concern I have is it might introduce a performance bottleneck and it's good to add a index over the tags in Lowercase.

Reply:
I think it's unlikely in practice to be run over a very large number of values, but if that use case comes up, you're right that an index might help.

I've found over the years that some of the most frustrating and challenging bugs come from issues related to character encoding and representation of time -- both things that seem like they should be simple, yet turn out not to be!